MF-397 - Introduce Thing Groups

mteodor opened the pull request:
Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

Adding group of things

mteodor committed:
add things group

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
add things group repository

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
add things group repository

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
add things group repository

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on things/api/logging.go:
```diff
@@ -264,3 +264,120 @@ func (lm *loggingMiddleware) Identify(ctx context.Context, key string) (id strin
+
+func (lm *loggingMiddleware) CreateGroup(ctx context.Context, token string, group things.Group) (g things.Group, err error) {
+	defer func(begin time.Time) {
+		message := fmt.Sprintf("Method create_group for key %s and name %s took %s to complete", token, g.Name(), time.Since(begin))
```
Here and for all methods: `key` or `token`?

manuio commented on things/api/things/http/transport.go:
```diff
@@ -172,6 +172,76 @@ func MakeHandler(tracer opentracing.Tracer, svc things.Service) http.Handler {
 		opts...,
 	))
 
+	r.Get("/things/:memberID/groups", kithttp.NewServer(
```
`thingID`

mteodor replied:
reason why it is `memberID` is because the methods that server this endpoint could be used for other entities that we want to group

manuio replied:
Until we decide let's have something consistent

mteodor replied:
we need to decide first

manuio replied:
What I meant is that if we finally decide to don't use generic groups you will have to remember and modify this. If we use generic groups this will moved from here in any case.

dborovcanin replied:
Since we have Groups API layer extracted to an internal package, I think that these remarks can be accepted.

manuio commented on things/api/things/http/transport.go:
```diff
@@ -172,6 +172,76 @@ func MakeHandler(tracer opentracing.Tracer, svc things.Service) http.Handler {
+		opts...,
+	))
+
+	r.Put("/groups/:groupID/things/:memberID", kithttp.NewServer(
```
`thingID`

manuio commented on things/api/things/http/transport.go:
```diff
@@ -172,6 +172,76 @@ func MakeHandler(tracer opentracing.Tracer, svc things.Service) http.Handler {
+		opts...,
+	))
+
+	r.Delete("/groups/:groupID/things/:memberID", kithttp.NewServer(
```
`thingID`

mteodor committed:
add parents and children methods

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
add parents and children methods

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

drasko commented on things/api/metrics.go:
```diff
@@ -192,3 +192,99 @@ func (ms *metricsMiddleware) Identify(ctx context.Context, key string) (string,
+func (ms *metricsMiddleware) Memberships(ctx context.Context, token, groupID string, offset, limit uint64, meta things.Metadata) (gp things.GroupPage, err error) {
+	defer func(begin time.Time) {
+		ms.counter.With("method", "memberships").Add(1)
+		ms.latency.With("method", "membeships").Observe(time.Since(begin).Seconds())
```
Fix typo

drasko commented on things/groups.go:
```diff
@@ -0,0 +1,168 @@
+	Description() string
+	Metadata() Metadata
+
+	SetID(id string)
```
Do we really need all these setters? Why not passing group struct pointer and changing the value directly.

This looks like some OO encapsulation, maybe it is needed but it Go it gives me "code smell".

mteodor replied:
Idea here and in `things/http.go` is to extract the common code that can be used for other entities that we wish to group
so `Group` is an interface that would have its implementations in each entity package itself
Of course if this solution is ok, http.go and `groups.go` would be moved to appropriate package

drasko replied:
This is OK, but I do not like these setters. You can create an var struct without the setters, and then change internal struct directly through pointer receiver.

manuio commented on things/api/logging.go:
```diff
@@ -264,3 +264,146 @@ func (lm *loggingMiddleware) Identify(ctx context.Context, key string) (id strin
+
+func (lm *loggingMiddleware) CreateGroup(ctx context.Context, token string, group things.Group) (g things.Group, err error) {
+	defer func(begin time.Time) {
+		message := fmt.Sprintf("Method create_group for key %s and name %s took %s to complete", token, group.Name(), time.Since(begin))
```
key or token?

mteodor committed:
fix logging message and temporary test fix

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix bootstrap test fail

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

dborovcanin commented on things/api/logging.go:
```diff
@@ -264,3 +264,146 @@ func (lm *loggingMiddleware) Identify(ctx context.Context, key string) (id strin
+	return lm.svc.Group(ctx, token, id)
+}
+
+func (lm *loggingMiddleware) Groups(ctx context.Context, token string, offset, limit uint64, meta things.Metadata) (gp things.GroupPage, err error) {
```
We usually use `ListEntity` naming for listing.

dborovcanin commented on things/api/metrics.go:
```diff
@@ -192,3 +192,99 @@ func (ms *metricsMiddleware) Identify(ctx context.Context, key string) (string,
+	return ms.svc.Group(ctx, token, id)
+}
+
+func (ms *metricsMiddleware) Groups(ctx context.Context, token string, offset, limit uint64, meta things.Metadata) (gp things.GroupPage, err error) {
```
Same, `ListGroups`.

dborovcanin commented on things/groups.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+type GroupRepository interface {
+
```
Please remove an empty line.

dborovcanin commented on things/groups.go:
```diff
@@ -0,0 +1,168 @@
+	"context"
+)
+
+type Group interface {
```
I'm also not sure about this idea. Why can't we use Group struct here? This interface resembles struct anyways (explains fields, rather than behavior). It is a limitation compared to the interface, but it also provides better type safety and this looks sufficient for Mainflux needs anyways.

mteodor replied:
Because Thing groups and User groups could differ in implementation and this interface should be common denominator.
Idea was to have at least some parts modular, that I put this into a package and reuse it in Users, Things and Channels groups, no need to rewrite again each endpoint for example.
This is something that was asked for by @drasko, at least one part. I can easily copy paste and refactor this into the usual place

drasko replied:
@mteodor this is OK, but remove getters and setters, use strucs and change the fields directly via pointer receiver.

dborovcanin replied:
Group entity itself, however, is the same in both cases, right?

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,601 @@
+	qd := `DELETE FROM thing_groups WHERE id = :id`
+	group := things.NewGroup()
+	group.SetID(groupID)
+	dbg, err := toDBGroup(group)
```
Can you simply create `dbGroup{ID: groupID}`?

mteodor committed:
move groups to pkg

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
move groups to pkg

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
move groups to pkg

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
move groups to internal

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
move groups to internal

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix import

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix linter errors

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

drasko commented on things/service.go:
```diff
@@ -382,3 +396,123 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+	}
+	group.ID = uid
+
+	// Temporary solution, until we have ID in response
```
Make this comment single line.

drasko commented on things/tracing/groups.go:
```diff
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: Apache-2.0
+
+// Package tracing contains middlewares that will add spans
+// to existing traces.
```
No need to break line here

mteodor committed:
fix comments

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on internal/groups/errors.go:
```diff
@@ -0,0 +1,14 @@
+	ErrDeleteGroupMissing = errors.New("cannot delete group")
+	ErrNotFound           = errors.New("cannot find group")
+	ErrAssignToGroup      = errors.New("cannot assign member to a group")
+	ErrConflict           = errors.New("group conflict")
```
Is this one duplicated of `ErrGroupConflict` ?

manuio commented on internal/groups/errors.go:
```diff
@@ -0,0 +1,14 @@
+	ErrCreateGroup        = errors.New("cannot create group")
+	ErrDeleteGroupMissing = errors.New("cannot delete group")
+	ErrNotFound           = errors.New("cannot find group")
+	ErrAssignToGroup      = errors.New("cannot assign member to a group")
```
Can we maybe use `failed to create, delete, find, assign member to group` ?

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,109 @@
+	ViewGroup(ctx context.Context, token, id string) (Group, error)
+
+	// ListGroups retrieves groups.
+	ListGroups(ctx context.Context, token string, offset, limit uint64, meta Metadata) (GroupPage, error)
```
Use `m Metadata` please

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,109 @@
+	RetrieveByName(ctx context.Context, name string) (Group, error)
+
+	// RetrieveAll retrieves all groups.
+	RetrieveAll(ctx context.Context, offset, limit uint64, gm Metadata) (GroupPage, error)
```
Same here `m Metadata`

manuio commented on things/api/logging.go:
```diff
@@ -264,3 +265,146 @@ func (lm *loggingMiddleware) Identify(ctx context.Context, key string) (id strin
 	return lm.svc.Identify(ctx, key)
 }
+
+func (lm *loggingMiddleware) CreateGroup(ctx context.Context, token string, group groups.Group) (g groups.Group, err error) {
```
Use `g groups.Group` please

manuio commented on things/api/logging.go:
```diff
@@ -264,3 +265,146 @@ func (lm *loggingMiddleware) Identify(ctx context.Context, key string) (id strin
+	return lm.svc.CreateGroup(ctx, token, group)
+}
+
+func (lm *loggingMiddleware) UpdateGroup(ctx context.Context, token string, group groups.Group) (g groups.Group, err error) {
```
same here

manuio commented on things/postgres/things.go:
```diff
@@ -22,6 +22,14 @@ const (
+	errUpdateDB   = errors.New("Update to DB failed")
+	errRetrieveDB = errors.New("Retreiving from DB failed")
+	errMarshal    = errors.New("Failed to marshal metadata")
+	errUnmarshal  = errors.New("Failed to unmarshal metadata")
```
Use lower case for error messages. You can also improve the consistency starting messages by `failed to ...`

manuio commented on things/service.go:
```diff
@@ -36,6 +40,12 @@ var (
+	ErrCreateGroup = errors.New("failed to create group")
+
+	// ErrFailedToRetrieveThings failed to retrieve things.
+	ErrFailedToRetrieveThings = errors.New("failed to retrieve things for group")
```
Maybe better `failed to retrieve group things` or `failed to retrieve group members`

mteodor committed:
resolve comments, add hierarchy info when retrieving groups

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

dborovcanin commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,111 @@
+	ListChildren(ctx context.Context, token, parentID string, offset, limit uint64, m Metadata) (GroupPage, error)
+
+	// ListParents retrieves groups that are parent to group identified by childID.
+	ListParents(ctx context.Context, token, childID string, offset, limit uint64, m Metadata) (GroupPage, error)
```
While I understand that using `parents` and `children` terminology is natural, I'm afraid it's also ambiguous because `parents` imply that one Group has multiple parents, and `children` implies only direct descendants.

dborovcanin replied:
Also, did you consider adding a feature to search genealogy only up to some depth, something like: give me only my children? Or: give me my only my children and children of my children. Same for parents. There are many possibilities for searching group members, but let's not get into that in this PR, this is just a remark for future reference.

dborovcanin replied:
Regarding https://github.com/mainflux/mainflux/pull/1259/files#r511910630, do Service methods return direct ascendents and descendants or the entire tree? This should be better explained in the documentation comments so that the one who implements knows the desirable behavior of the `Service` interface.

dborovcanin commented on internal/groups/http.go:
```diff
@@ -0,0 +1,685 @@
+	}
+}
+
+func AssignMemberToGroup(svc Service) endpoint.Endpoint {
```
Please use consistent naming. We usually use the `Endpoint` suffix. If the method name is too long, you can shorten it like `AssignEndpoint` or `AssignMemberEndpoint`, where `member` implies that we're talking about Groups.

dborovcanin commented on internal/groups/http.go:
```diff
@@ -0,0 +1,685 @@
+	}
+}
+
+func RemoveMemberFromGroup(svc Service) endpoint.Endpoint {
```
Same here: `UnassignEndpoint` or `UnassignMemberEndpoint`.

dborovcanin commented on internal/groups/http.go:
```diff
@@ -0,0 +1,685 @@
+	}
+}
+
+func ListMembersForGroupEndpoint(svc Service) endpoint.Endpoint {
```
Same here: Member implies Group, so `ListMembersEndpoint` is sufficient.

dborovcanin commented on internal/groups/http.go:
```diff
@@ -0,0 +1,685 @@
+package groups
```
Consider moving this to a separate `http` or `api` package (`/internal/groups/http` or `/internal/groups/api`) and breaking this file to the usual Mainflux API package structure (endpoints.go, requests.go, responses.go...).

dborovcanin commented on internal/groups/http.go:
```diff
@@ -0,0 +1,685 @@
+	}
+}
+
+func DecodeListGroupsRequest(_ context.Context, r *http.Request) (interface{}, error) {
```
Consider separating decode methods from endpoint methods (break in two files `endpoint.go` and `transport.go`).

dborovcanin commented on internal/groups/http.go:
```diff
@@ -0,0 +1,685 @@
+	return res
+}
+
+// responses.go
```
Exactly, please move this to a new file.

dborovcanin commented on internal/groups/http.go:
```diff
@@ -0,0 +1,685 @@
+	return true
+}
+
+// requests.go
```
Same.

dborovcanin commented on pkg/sdk/go/things_test.go:
```diff
@@ -49,7 +49,7 @@ func newThingsService(tokens map[string]string) things.Service {
 	uuidProvider := uuid.NewMock()
 
-	return things.New(auth, thingsRepo, channelsRepo, chanCache, thingCache, uuidProvider)
+	return things.New(auth, thingsRepo, channelsRepo, nil, chanCache, thingCache, uuidProvider)
```
Please open an issue for SDK support for Things groups.

dborovcanin replied:
Issue #1270 (CLI is using SDK).

dborovcanin commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,111 @@
+
+type MemberPage struct {
+	PageMetadata
+	Members []Member
```
Do we need Member type definition or `interface{}` is sufficient? I have no preferences, but we probably don't gain anything with this since we need to typecast anyways.

dborovcanin commented on things/api/logging.go:
```diff
@@ -264,3 +265,146 @@ func (lm *loggingMiddleware) Identify(ctx context.Context, key string) (id strin
 	return lm.svc.Identify(ctx, key)
 }
+
+func (lm *loggingMiddleware) CreateGroup(ctx context.Context, token string, g groups.Group) (gr groups.Group, err error) {
```
Please be consistent when using `token` and `key`. I'm OK with using either option. It's a JSON Web **Token** which is Mainflux **Key**, but, using it as `key` here and as `token` there might be interpreted as if it's not the same thing.

dborovcanin commented on things/postgres/init.go:
```diff
@@ -93,6 +93,29 @@ func migrateDB(db *sqlx.DB) error {
+						description VARCHAR(1024),
+						metadata    JSONB,
+						PRIMARY KEY (id),
+						FOREIGN KEY (parent_id) REFERENCES thing_groups (id)  ON DELETE CASCADE ON UPDATE CASCADE
```
Please check the formatting, it looks like there are multiple whitespaces in some lines that start with `FOREIGN_KEY`.

dborovcanin commented on things/service.go:
```diff
@@ -382,3 +396,122 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+	group.ID = uid
+
+	// Temporary solution, until we have ID in response from Identify gRPC call.
+	owner, err := uuidProvider.New().ID()
```
We need to prioritize #1245 and merge it before this PR.

mteodor committed:
code refactor, separate http into multiple files

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
code refactor, separate http into multiple files

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix group update

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
use user id when saving

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
use user id when saving

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
rename methods

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
move code

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
move code

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
remove temporary test

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
add groups test

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,252 @@
+			ParentID:    gp.ParentID,
+			Description: gp.Description,
+			Metadata:    gp.Metadata,
+			Name:        gp.Name,
```
Only `ID` is needed for the "location header". Remove other fields please

mteodor committed:
fix comments, fix responses in api for groups

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
revert changes

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

add checks for name length

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

fix validation

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

small change to response

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

add comment for Level

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

add level to limit hierarchy retrieval

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

tidy vendor

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

add level to limit hierarchy retrieval

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mod tidy

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

revert

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

resolve comments

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

small naming and code organize refactor

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

revert Member type

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

fix typo

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

use ltree

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

use ltree

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

add level for retrieving

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

add level

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

add ltre

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

upgrade postgres version in test

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

add ltre

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

remove test for now

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

minor fixes

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

fix id setting

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

fix tree endpoint

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

minor style changes

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,241 @@
+	}
+}
+
+// endpoint.go
```
What is this comment?

manuio commented on internal/groups/api/requests.go:
```diff
@@ -0,0 +1,109 @@
+	if req.ID == "" {
+		return groups.ErrMalformedEntity
+	}
+
```
check name size here

dborovcanin commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,241 @@
+	defLimit  = 10
+)
+
+func ListMembership(svc groups.Service) endpoint.Endpoint {
```
Can you organize methods in the usual CRUD order: create, read(view, list), update, delete? It's easier to read the file if methods are in the expected order.

dborovcanin commented on internal/groups/errors.go:
```diff
@@ -0,0 +1,41 @@
+package groups
```
I like that errors are extracted to a separate file.

dborovcanin commented on things/postgres/channels.go:
```diff
@@ -453,18 +453,17 @@ func getMetadataQuery(m things.Metadata) ([]byte, string, error) {
 		return 0, err
 	}
-
+	defer rows.Close()
```
Good catch.

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,111 @@
+	RetrieveAll(ctx context.Context, offset, limit uint64, gm Metadata) (GroupPage, error)
+
+	// RetrieveAllParents retrieves all groups that are ancestors to the group with given groupID.
+	RetrieveAllParents(ctx context.Context, groupID string, offset, limit uint64, gm Metadata) (GroupPage, error)
```
`m Metdata`

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+type Repository interface {
+
```
Remove this blank line

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,111 @@
+	RetrieveAllParents(ctx context.Context, groupID string, offset, limit uint64, gm Metadata) (GroupPage, error)
+
+	// RetrieveAllChildren retrieves all children from group with given groupID.
+	RetrieveAllChildren(ctx context.Context, groupID string, offset, limit uint64, gm Metadata) (GroupPage, error)
```
same here

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,111 @@
+	RetrieveAllChildren(ctx context.Context, groupID string, offset, limit uint64, gm Metadata) (GroupPage, error)
+
+	// Retrieves list of groups that member belongs to
+	Memberships(ctx context.Context, memberID string, offset, limit uint64, gm Metadata) (GroupPage, error)
```
and here

dborovcanin commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,111 @@
+	Description string
+	Metadata    Metadata
+	Level       int
+	Path        string
```
I'm not sure that these belong to the Group struct. If we need info about the position of the Group in the hierarchy, we can expose an API for that. I doubt that we'll need it. Also, please consider limiting the possible depth of this hierarchy as well as the number of direct children to some sane number (10 or 20 should be enough) due to the performance constraints.

drasko commented on bootstrap/mocks/things.go:
```diff
@@ -230,3 +231,47 @@ func findIndex(list []string, val string) int {
 }
+
+func (svc *mainfluxThings) CreateGroup(ctx context.Context, token string, group groups.Group) (groups.Group, error) {
+	panic("not implemented")
```
No need to panic and break the whole service

drasko replied:
Also, please open an issue for this and paste the issue reference as a comment to this PR.

mteodor replied:
This is a mock, and while it isn't used it will not break anything similar to other places

dborovcanin replied:
Comment resolved by issue #1271.

drasko commented on internal/groups/api/responses.go:
```diff
@@ -0,0 +1,146 @@
+	OwnerID     string                 `json:"owner_id,omitempty"`
+	Description string                 `json:"description,omitempty"`
+	Metadata    map[string]interface{} `json:"metadata,omitempty"`
+	Level       int                    `json:"level,omitempty"`
```
What is this level? It is unclear from the code reader's perspective and probably demands a comment.

dborovcanin commented on things/postgres/groups_test.go:
```diff
@@ -0,0 +1,490 @@
+	}
+}
+
+// func TestAssignUser(t *testing.T) {
```
Please remove the commented code.

dborovcanin commented on things/postgres/groups_test.go:
```diff
@@ -0,0 +1,490 @@
+	"github.com/stretchr/testify/require"
+)
+
+/*
```
Please remove the commented code.

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,631 @@
+	Size     uint64
+}
+
+func toUUID(id string) (uuid.NullUUID, error) {
```
This looks way too complicated. Can you use `uuid.FromStringOrNil(id)` or at least simplify this method?

dborovcanin replied:
This whole Null UUID handling looks complicated. Can you try to simplify it?

mteodor replied:
I already did simplification and this best I can do

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,631 @@
+		mq = fmt.Sprintf("WHERE %s", mq)
+	}
+
+	sq := `WITH RECURSIVE subordinates (id, owner_id, parent_id, name, description, metadata,  level, path) AS (
```
This query is complex and it hard to read and understand. A large part of complexity comes from having `path` and `level`. If we remove it, do we still have all the necessary data to reconstruct Group genealogy? I mean, can we construct a tree or whichever data structure we want by programmatically tracking IDs of the retrieved groups?

dborovcanin commented on internal/groups/api/requests.go:
```diff
@@ -0,0 +1,117 @@
+	if req.token == "" {
+		return groups.ErrUnauthorizedAccess
+	}
+	// if req.groupID == "" && req.memberID == "" {
```
Please remove the commented code.

dborovcanin commented on internal/groups/api/requests.go:
```diff
@@ -0,0 +1,117 @@
+	if req.token == "" {
+		return groups.ErrUnauthorizedAccess
+	}
+	// // if req.groupID == "" && req.name == "" {
```
Please remove the commented code.

dborovcanin commented on internal/groups/api/requests.go:
```diff
@@ -0,0 +1,117 @@
+	if req.token == "" {
+		return groups.ErrUnauthorizedAccess
+	}
+	// if req.groupID == "" && req.memberID == "" {
```
Please remove the commented code.

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,576 @@
+	if group.Description == "" {
+		group.Description = old.Description
+	}
+	// if group.ParentID == "" {
```
Please remove the commented code.

manuio commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,245 @@
+)
+
+const (
+	maxNameSize = 256
```
Considering that name is `VARCHAR(254)`  is this 256 correct?

manuio commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,245 @@
+	contentType = "application/json"
+
+	defOffset = 0
+	defLimit  = 10
```
This consts should be defined in transport.go

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,122 @@
+	Total  uint64
+	Offset uint64
+	Limit  uint64
+	Name   string
```
Do we need a name for the page meta?

mteodor replied:
Yes I think, it is useful for example when returning members, since members can be either things or users, so name is additional info

drasko replied:
Can you give an example please, I am not convinced with this explanation.

dborovcanin replied:
I'm also not convinced. You can't confuse Things and Users Groups because the Members will be of different types.

mteodor replied:
```
curl -s -S -X GET http://localhost:9001/groups/894f95a2-14b9-4f4b-9164-ff39a5ee7114/children\?level\=5\&tree\=true  -H "Authorization: $TOK" -H 'Content-Type: application/json' | jq                                                                    
{                                                                                                                                                                                                                                                                                         
  "total": 3,                                                                                                                                                                                                                                                                             
  "offset": 0,                                                                                                                                                                                                                                                                            
  "limit": 0,                                                                                                                                                                                                                                                                             
  "name": "things",                                                                                                                                                                                                                                                                             
  "groups": [                                                                                                                                                                                                                                                                             
    {                                                                                                                                                                                                                                                                                     
      "id": "894f95a2-14b9-4f4b-9164-ff39a5ee7114",                                                                                                                                                                                                                                       
      "name": "group6",                                                                                                                                                                                                                                                                   
      "owner_id": "d3861ca6-a950-4a51-91ff-50cfb5559a11",                                                                                                                                                                                                                                 
      "metadata": {                                                                                                                                                                                                                                                                       
        "admin": "true"                                                                                                                                                                                                                                                                   
      },                                                                                                                                                                                                                                                                                  
      "level": 1,                                                                                                                                                                                                                                                                         
      "path": "group6",                                                                                                                                                                                                                                                                   
      "children": [               
```


and url would be similar for users just different port

dborovcanin replied:
Thanks for the explanation. I'm OK with this.

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,122 @@
+	CreateGroup(ctx context.Context, token string, group Group) (Group, error)
+
+	// UpdateGroup updates the group identified by the provided ID.
+	UpdateGroup(ctx context.Context, token string, group Group) (Group, error)
```
`g Group`

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,122 @@
+	Save(ctx context.Context, group Group) (Group, error)
+
+	// Update a group
+	Update(ctx context.Context, group Group) (Group, error)
```
`g Group`

manuio commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,122 @@
+	RetrieveAllChildren(ctx context.Context, groupID uuid.NullUUID, offset, limit uint64, gm Metadata) (GroupPage, error)
+
+	// Retrieves list of groups that member belongs to
+	Memberships(ctx context.Context, memberID uuid.NullUUID, offset, limit uint64, gm Metadata) (GroupPage, error)
```
`m Metadata`

manuio commented on things/api/things/http/transport.go:
```diff
@@ -172,6 +174,83 @@ func MakeHandler(tracer opentracing.Tracer, svc things.Service) http.Handler {
+	))
+
+	r.Get("/groups", kithttp.NewServer(
+		kitot.TraceServer(tracer, "groups")(groupsAPI.ListGroupsEndpoint(svc)),
```
`list_groups`

manuio commented on things/api/things/http/transport.go:
```diff
@@ -172,6 +174,83 @@ func MakeHandler(tracer opentracing.Tracer, svc things.Service) http.Handler {
+	))
+
+	r.Put("/groups/:groupID/things/:memberID", kithttp.NewServer(
+		kitot.TraceServer(tracer, "assign_user_to_group")(groupsAPI.AssignEndpoint(svc)),
```
`assign`

manuio commented on things/api/things/http/transport.go:
```diff
@@ -172,6 +174,83 @@ func MakeHandler(tracer opentracing.Tracer, svc things.Service) http.Handler {
+	))
+
+	r.Delete("/groups/:groupID/things/:memberID", kithttp.NewServer(
+		kitot.TraceServer(tracer, "remove_thing_from_group")(groupsAPI.UnassignEndpoint(svc)),
```
unassign

manuio commented on things/api/things/http/transport.go:
```diff
@@ -172,6 +174,83 @@ func MakeHandler(tracer opentracing.Tracer, svc things.Service) http.Handler {
+	))
+
+	r.Get("/groups/:groupID/things", kithttp.NewServer(
+		kitot.TraceServer(tracer, "members")(groupsAPI.ListMembersEndpoint(svc)),
```
list_members

manuio commented on things/api/things/http/transport.go:
```diff
@@ -172,6 +174,83 @@ func MakeHandler(tracer opentracing.Tracer, svc things.Service) http.Handler {
+	))
+
+	r.Get("/groups/:groupID", kithttp.NewServer(
+		kitot.TraceServer(tracer, "group")(groupsAPI.ViewGroupEndpoint(svc)),
```
`view_group`

manuio commented on things/service.go:
```diff
@@ -382,3 +397,118 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
 	return thingID, nil
 }
+
+func (ts *thingsService) CreateGroup(ctx context.Context, token string, group groups.Group) (groups.Group, error) {
```
`g groups.Group`

manuio commented on things/service.go:
```diff
@@ -382,3 +397,118 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+	return ts.groups.Save(ctx, group)
+}
+
+func (ts *thingsService) ListGroups(ctx context.Context, token string, offset, limit uint64, meta groups.Metadata) (groups.GroupPage, error) {
```
`gm groups.Metadata`

manuio commented on things/service.go:
```diff
@@ -382,3 +397,118 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+
+}
+
+func (ts *thingsService) ListParents(ctx context.Context, token string, childID uuid.NullUUID, offset, limit uint64, meta groups.Metadata) (groups.GroupPage, error) {
```
`gm groups.Metadata`

manuio commented on things/service.go:
```diff
@@ -382,3 +397,118 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+	return ts.groups.RetrieveAllParents(ctx, childID, offset, limit, meta)
+}
+
+func (ts *thingsService) ListChildren(ctx context.Context, token string, parentID uuid.NullUUID, offset, limit uint64, meta groups.Metadata) (groups.GroupPage, error) {
```
`gm groups.Metadata`

manuio commented on things/tracing/groups.go:
```diff
@@ -0,0 +1,137 @@
+	saveGroup            = "save_group"
+	deleteGroup          = "delete_group"
+	updateGroup          = "update_group"
+	retrieveGroupByID    = "retrieve_group_by_id"
```
`	retrieveByID    = "retrieve_by_id"`

manuio commented on things/tracing/groups.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+// GroupRepositoryMiddleware tracks request and their latency, and adds spans to context.
+func GroupRepositoryMiddleware(tracer opentracing.Tracer, repo groups.Repository) groups.Repository {
```
`gr groups.Repository`

manuio commented on things/tracing/groups.go:
```diff
@@ -0,0 +1,137 @@
+		repo:   repo,
+	}
+}
+func (grm groupRepositoryMiddleware) Save(ctx context.Context, group groups.Group) (groups.Group, error) {
```
`g groups.Group`

manuio commented on things/tracing/groups.go:
```diff
@@ -0,0 +1,137 @@
+
+	return grm.repo.Save(ctx, group)
+}
+func (grm groupRepositoryMiddleware) Update(ctx context.Context, group groups.Group) (groups.Group, error) {
```
`g groups.Group`

manuio commented on internal/groups/api/requests.go:
```diff
@@ -0,0 +1,126 @@
+
+type updateGroupReq struct {
+	token       string
+	ID          string                 `json:"id,omitempty"`
```
This field don't need to be exported. It comes as path parameter not from the body of the request.

manuio commented on internal/groups/api/transport.go:
```diff
@@ -0,0 +1,201 @@
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return nil, errors.Wrap(groups.ErrFailedDecode, err)
+	}
+	req.ID = bone.GetValue(r, "groupID")
```
Here it is

manuio commented on users/service.go:
```diff
@@ -330,23 +330,18 @@ func (svc usersService) CreateGroup(ctx context.Context, token string, group Gro
 	}
 
-	email, err := svc.identify(ctx, token)
+	id, err := svc.identify(ctx, token)
```
AFAIK `svc.identify` returns an `email`

manuio commented on things/tracing/groups.go:
```diff
@@ -0,0 +1,138 @@
+)
+
+const (
+	assignMember         = "assign_member"
```
This could be `assign`

manuio commented on things/tracing/groups.go:
```diff
@@ -0,0 +1,138 @@
+	retrieveByName       = "retrieve_by_name"
+	memberships          = "memberships"
+	members              = "members"
+	unassignMember       = "unassign_member"
```
This could be `unassign`

manuio commented on things/service.go:
```diff
@@ -382,3 +395,115 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+			Total:  p.Total,
+			Offset: p.Offset,
+			Limit:  p.Limit,
+			Name:   "things",
```
Use a const for `"things"`

manuio commented on things/postgres/things.go:
```diff
@@ -22,6 +22,11 @@ const (
 
+var (
+	errUpdateDB   = errors.New("failed to update DB")
+	errRetrieveDB = errors.New("failed retrieving from DB")
```
not sure about capital letters in logs

manuio commented on things/postgres/init.go:
```diff
@@ -93,6 +93,31 @@ func migrateDB(db *sqlx.DB) error {
+						name        VARCHAR(254) UNIQUE NOT NULL,
+						description VARCHAR(1024),
+						metadata    JSONB,
+						path 		LTREE, 
```
Looks like tabulation is not correct here

manuio commented on things/postgres/groups_test.go:
```diff
@@ -0,0 +1,309 @@
+			Name:        fmt.Sprintf("mainflux%d", i),
+		}
+
+		// Create Things with metadata.
```
`Things` or `Groups`?

manuio commented on things/postgres/groups_test.go:
```diff
@@ -0,0 +1,309 @@
+			err:     nil,
+		},
+		{
+			desc:    "retrieve a group with incorrect id",
```
Maybe `invalid` instead of `incorrect`

manuio commented on things/postgres/groups_test.go:
```diff
@@ -0,0 +1,309 @@
+		err     error
+	}{
+		{
+			desc:    "retrieve a group",
```
`retrieve group with valid id`

manuio commented on things/postgres/groups_test.go:
```diff
@@ -0,0 +1,309 @@
+			err:   nil,
+		},
+		{
+			desc:  "create group with invalid parentID",
```
`invalid parent id`

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,308 @@
+			return groupRes{}, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+
+		group := groups.Group{
```
Here we use `group` and in the function before `g`. Please choose one.

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,308 @@
+		if err := req.validate(); err != nil {
+			return groupPageRes{}, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+		gp, err := svc.ListGroups(ctx, req.token, req.level, req.metadata)
```
Here again is `gp`, not even aligned to any of previous two :)

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,308 @@
+		if err := req.validate(); err != nil {
+			return memberPageRes{}, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+		mp, err := svc.ListMembers(ctx, req.token, req.groupID, req.offset, req.limit, req.metadata)
```
What is `mp`? I would understand `m`, but what is this `p`?

mteodor replied:
Member page

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,308 @@
+	groupsMap := map[string]*groups.Group{}
+	// Parents map keeps its array of children.
+	parentsMap := map[string][]*groups.Group{}
+	for _, gr := range page.Groups {
```
Here yous use `gr`

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,308 @@
+		}
+	}
+
+	for _, g := range groupsMap {
```
And here you use `g`

drasko commented on internal/groups/api/transport.go:
```diff
@@ -0,0 +1,221 @@
+	if !strings.Contains(r.Header.Get("Content-Type"), contentType) {
+		return nil, groups.ErrUnsupportedContentType
+	}
+	o, err := readUintQuery(r, offsetKey, defOffset)
```
Add empty line here, if you added them below.

drasko commented on internal/groups/api/transport.go:
```diff
@@ -0,0 +1,221 @@
+func DecodeGroupCreate(_ context.Context, r *http.Request) (interface{}, error) {
+	if !strings.Contains(r.Header.Get("Content-Type"), contentType) {
+		return nil, groups.ErrUnsupportedContentType
+	}
```
Add empty line below.

drasko commented on internal/groups/api/transport.go:
```diff
@@ -0,0 +1,221 @@
+	var req createGroupReq
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return nil, errors.Wrap(groups.ErrFailedDecode, err)
+	}
```
Add empty line below.

drasko commented on internal/groups/api/transport.go:
```diff
@@ -0,0 +1,221 @@
+	if !strings.Contains(r.Header.Get("Content-Type"), contentType) {
+		return nil, groups.ErrUnsupportedContentType
+	}
+	var req updateGroupReq
```
Same here

drasko commented on things/service.go:
```diff
@@ -382,3 +396,115 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+}
+
+func (ts *thingsService) ListGroups(ctx context.Context, token string, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	_, err := ts.auth.Identify(ctx, &mainflux.Token{Value: token})
```
If value is ignored, embed the check.

drasko commented on things/service.go:
```diff
@@ -382,3 +396,115 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+}
+
+func (ts *thingsService) ListParents(ctx context.Context, token string, childID string, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	_, err := ts.auth.Identify(ctx, &mainflux.Token{Value: token})
```
Same here

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,316 @@
+			Metadata:    req.Metadata,
+		}
+
+		g, err := svc.CreateGroup(ctx, req.token, group)
```
It is unclear from the code (on reading) what is the difference between `group` and `g` - they both suggest to be groups. It is also expected for `CreateGroup()` to return group, but on the other hand you created it manually above.

mteodor replied:
same as `CreateThings` works

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,316 @@
+		if err := req.validate(); err != nil {
+			return groupPageRes{}, errors.Wrap(groups.ErrMalformedEntity, err)
+		}
+		page, err := svc.ListParents(ctx, req.token, req.groupID, req.level, req.metadata)
```
Please stay consistent in adding empty lines comparing to other function.

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,316 @@
+	// Parents map keeps its array of children.
+	parentsMap := map[string][]*groups.Group{}
+	for _, g := range page.Groups {
+		group := g
```
Why is this `group` var needed? Can't just `g` be used here directly?

mteodor replied:
no it cant, it creates new object as a copy of `g` so that its address can be used for pointer in next lines

manuio replied:
Probably this would solve your issue:
```
	for i := range page.Groups {
		if _, ok := groupsMap[group.ID]; !ok {
			groupsMap[group.ID] = &page.Groups[i]
			parentsMap[group.ID] = make([]*groups.Group, 0)
		}
	}
```

drasko commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,316 @@
+		child := toViewGroupRes(*ch)
+		view.Children = append(view.Children, &child)
+	}
+	return view
```
If empty line was added above, add one here as well

drasko commented on internal/groups/api/requests.go:
```diff
@@ -0,0 +1,126 @@
+	if req.token == "" {
+		return groups.ErrUnauthorizedAccess
+	}
+
```
If function below do not have white lines, please suppers these as well.

manuio replied:
To be consistent I'd add it in the function where it's missing.

mteodor committed:
small changes, adding new lines

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,336 @@
+		if ch, ok := parentsMap[g.ParentID]; ok {
+			ch = append(ch, g)
+			parentsMap[g.ParentID] = ch
+			fmt.Printf("%s, array> %d\n", g.ParentID, len(ch))
```
Remove this log

mteodor committed:
minor changes

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on internal/groups/api/endpoint.go:
```diff
@@ -0,0 +1,335 @@
+			Metadata:    req.Metadata,
+		}
+
+		created, err := svc.CreateGroup(ctx, req.token, group)
```
`created` should be called `group` but in fact `svc.CreateGroup` should return `groupID` instead of whole Group.

dborovcanin commented on authn/postgres/setup_test.go:
```diff
@@ -32,7 +32,7 @@ func TestMain(m *testing.M) {
 		"POSTGRES_DB=test",
 	}
-	container, err := pool.Run("postgres", "10.2-alpine", cfg)
+	container, err := pool.Run("postgres", "10.8-alpine", cfg)
```
Consider updating version in correpsonding `docker-compose.ylm` files.

dborovcanin commented on internal/groups/api/responses.go:
```diff
@@ -0,0 +1,151 @@
+	// Level if retrieved with Parents or Children method
+	// indicates a level in hierarchy from first group node.
+	Level    int             `json:"level,omitempty"`
+	Path     string          `json:"path"`
```
A short explanation similar to the one for `level` field would be nice. Also, is level empty (thus, omitted) if it's not Parents/Children method?

mteodor replied:
path and level are always there

dborovcanin commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,112 @@
+	Description string
+	Metadata    Metadata
+	// Level if retrieved with Parents or Children method
+	// indicates a level in hierarchy from first group node.
```
Is level 0 when the Group is not fetched using Parents/Children? This can be confusing and misleading if not documented. Also, consider using a pointer (or at least setting it to a negative value) so that there is a difference between 0 level and level not fetched at all.

dborovcanin commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,112 @@
+	RetrieveByID(ctx context.Context, id string) (Group, error)
+
+	// RetrieveByName retrieves group by its name
+	RetrieveByName(ctx context.Context, name string) (Group, error)
```
This implies that `Name` is unique. Instead, please consider making (owner, name) pair unique. If two users, for example, group Things by the same criteria (say, spacial), it's very likely that there will be a name collision.

mteodor replied:
Name is unique

drasko replied:
@mteodor how do you guarantee name uniqueness in a distributed system?

dborovcanin replied:
Based on the other conversations and on the code in the implementation (`init.go`), the name doesn't seem to be unique. Should this method signature be updated?

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,647 @@
+	if g.Description == "" {
+		g.Description = old.Description
+	}
+	if g.ParentID == "" {
```
I'm not sure that we should allow parent change. While it's an extremely powerful feature, I find it rather dangerous (upcoming policies are going to be heavily affected). @mainflux/maintainers Opinions?

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,647 @@
+	}
+	if g.Description == "" {
+		g.Description = old.Description
+	}
```
Consider allowing `name` and `description` to be empty. While it's pointless to have the `name` field empty, it will eliminate this entire fetch from the DB, as well as potential transaction problem we have here (fetching and updating are executed in separate transactions) due to the lack of optimistic lock.

mteodor committed:
change function signature

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
change primary key

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
change function signature

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,647 @@
+	}
+
+	q := fmt.Sprintf("SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level FROM thing_groups WHERE path @> :path AND nlevel(:path) - nlevel(path) >= 0 AND nlevel(:path) - nlevel(path) <= :level %s", mq)
+	cq := fmt.Sprintf("SELECT COUNT(*) FROM thing_groups WHERE path @> :path AND nlevel(:path) - nlevel(path) >= 0 AND nlevel(:path) - nlevel(path) <= :level %s", mq)
```
Split this in multiple lines

manuio commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,647 @@
+	}
+
+	q := fmt.Sprintf("SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level FROM thing_groups WHERE path <@ :path AND nlevel(path) - nlevel(:path) <= :level %s", mq)
+	cq := fmt.Sprintf("SELECT COUNT(*) FROM thing_groups WHERE path <@ :path AND nlevel(path) - nlevel(:path) <= :level %s", mq)
```
same here and for all other queries:

manuio commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,647 @@
+
+	q := fmt.Sprintf(`SELECT th.id, th.name, th.key, th.metadata FROM things th, thing_group_relations g
+                      WHERE th.id = g.thing_id AND g.group_id = :group 
+                      %s ORDER BY id LIMIT :limit OFFSET :offset;`, mq)
```
The tabulation seems wrong here

manuio commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,647 @@
+	if mq != "" {
+		mq = fmt.Sprintf("AND %s", mq)
+	}
+	q := fmt.Sprintf(`SELECT g.id, g.owner_id, g.parent_id, g.name, g.description, g.metadata 
+					  FROM thing_group_relations gr, thing_groups g
+					  WHERE gr.group_id = g.id and gr.thing_id = :userID 
+		  			  %s ORDER BY id LIMIT :limit OFFSET :offset;`, mq)
+
```
same here

manuio commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,647 @@
+
+	cq := fmt.Sprintf(`SELECT COUNT(*) 
+					   FROM thing_group_relations gr, thing_groups g
+					   WHERE gr.group_id = g.id and gr.thing_id = :userID %s;`, mq)
```
and here

manuio commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,647 @@
+	mq := ""
+	mb := []byte("{}")
+	if len(m) > 0 {
+		mq = db + `.metadata @> :metadata`
```
If `db` is empty is `.metadata @> :metadata` going to work?

manuio commented on things/service.go:
```diff
@@ -382,3 +396,117 @@ func (ts *thingsService) hasThing(ctx context.Context, chanID, thingKey string)
+
+func (ts *thingsService) ListChildren(ctx context.Context, token string, parentID string, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
+	_, err := ts.auth.Identify(ctx, &mainflux.Token{Value: token})
+	if err != nil {
```
Please, inline this check here and for all other methods

mteodor committed:
simplufy code

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on things/postgres/groups_test.go:
```diff
@@ -0,0 +1,420 @@
+	cleanAll(t, dbMiddleware)
+
+	ownerID, err := uuidProvider.New().ID()
+	require.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
```
Can you please find/replace `got unexpected error` by `unexpected error`

mteodor committed:
remove groups test for now

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix tabulation

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix whitespace

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio commented on internal/groups/errors.go:
```diff
@@ -0,0 +1,49 @@
+	// ErrMissingParent indicates that parent cant be found
+	ErrMissingParent = errors.New("failed to retrieve parent")
+
+
```
Remove these blank lines

manuio commented on internal/groups/errors.go:
```diff
@@ -0,0 +1,49 @@
+	// ErrFailedDecode indicates failed to decode request body
+	ErrFailedDecode = errors.New("failed to decode request body")
+
+	// ErrMissingParent indicates that parent cant be found
```
`can't`

mteodor committed:
remove white space, fix grammar

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

drasko commented on internal/groups/api/responses.go:
```diff
@@ -0,0 +1,152 @@
+	Offset uint64 `json:"offset"`
+	Limit  uint64 `json:"limit"`
+	Name   string `json:"name"`
+}
```
Add one empty line after the struct definition

dborovcanin commented on things/postgres/init.go:
```diff
@@ -93,6 +93,32 @@ func migrateDB(db *sqlx.DB) error {
+						description VARCHAR(1024),
+						metadata    JSONB,
+						path        LTREE, 
+						PRIMARY KEY (name, path),
```
Does it make sense to have an ID that's not a primary key?

mteodor replied:
https://github.com/mainflux/mainflux/pull/1259#discussion_r525561057 Im talking about json response

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,636 @@
+}
+
+func (gr groupRepository) Update(ctx context.Context, g groups.Group) (groups.Group, error) {
+	q := `UPDATE thing_groups SET name = :name, description = :description, parent_id = :parent_id, metadata = :metadata  WHERE id = :id`
```
Is it safe to enable Parent update? I'm afraid that it might be dangerous, especially once we introduce policies attached to the groups.

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,636 @@
+	q := fmt.Sprintf(`SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level FROM thing_groups 
+					  WHERE path <@ :path AND nlevel(path) - nlevel(:path) <= :level %s`, mq)
+	cq := fmt.Sprintf("SELECT COUNT(*) FROM thing_groups WHERE path <@ :path AND nlevel(path) - nlevel(:path) <= :level %s", mq)
+	g, err := gr.RetrieveByID(ctx, groupID)
```
Looks like you are fetching the group only to fetch a path for the response. Can this call be avoided (it's a separate transaction - one more DB driver call; also, because not everything is done in the transactions and can change parent in the current implementation - I'm afraid that we need an optimistic lock here)?

mteodor committed:
fix sqls so that not additional retrieve is needed

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix sqls so that not additional retrieve is needed

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

fix primary key

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

remove retrive by name

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

drasko reviewed (dismissed):
LGTM

manuio reviewed (dismissed):
LGTM

mteodor committed:
remove name and parent update

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix cound sql

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
add line

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio reviewed (dismissed):
LGTM

dborovcanin commented on things/postgres/init.go:
```diff
@@ -93,6 +93,32 @@ func migrateDB(db *sqlx.DB) error {
+						PRIMARY KEY (owner_id, path),
+						FOREIGN KEY (parent_id) REFERENCES thing_groups (id) ON DELETE CASCADE ON UPDATE CASCADE
+				   )`,
+					`CREATE TABLE IF NOT EXISTS id (
```
Table name `id` is less than ideal, consider renaming to `members`.

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,626 @@
+	q := fmt.Sprintf(`SELECT g.id, g.name, g.owner_id, g.parent_id, g.description, g.metadata, g.path, nlevel(g.path) as level FROM 
+					 (SELECT thing_groups.path FROM thing_groups where id = :parent_id) parent LEFT JOIN
+					 (SELECT id, name, owner_id, parent_id, description,metadata, path FROM thing_groups) g 
+					  ON g.path @> parent.path AND nlevel(parent.path) - nlevel(g.path) >= 0 AND nlevel(parent.path) - nlevel(g.path) <= :level %s`, mq)
```
Is `g.path @> parent.path AND nlevel(parent.path) - nlevel(g.path) >= 0` necessary, or we can use only `g.path @> parent.path`?

dborovcanin commented on things/postgres/groups.go:
```diff
@@ -0,0 +1,626 @@
+		mq = fmt.Sprintf("AND %s", mq)
+	}
+
+	q := fmt.Sprintf(`SELECT g.id, g.name, g.owner_id, g.parent_id, g.description, g.metadata, g.path, nlevel(g.path) as level FROM 
```
Consider using simpler `self join`. You can try this query:
```sql
SELECT g.id, g.name, g.owner_id, g.parent_id, g.description, g.metadata, g.path, nlevel(g.path) as level
FROM thing_groups parent, thing_groups g
WHERE parent.id = :id AND g.path @> parent.path AND nlevel(parent.path) - nlevel(g.path) <= :level;
```
It's a little simpler (and shorter) and possibly more performant.

mteodor committed:
simplify sql, fix table name

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

manuio reviewed (dismissed):
LGTM

mteodor committed:
add date

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
add date

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

drasko commented on internal/groups/api/responses.go:
```diff
@@ -0,0 +1,153 @@
+
+type groupRes struct {
+	id      string
+	created bool
```
Missing `updated`?

drasko replied:
`created` is a time, not bool. Time of creation.

mteodor committed:
add timestamp to api

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

dborovcanin commented on internal/groups/groups.go:
```diff
@@ -0,0 +1,114 @@
+	Path      string
+	Children  []*Group
+	CreatedAt time.Time
+	UpdateAt  time.Time
```
`UpdatedAt`

mteodor committed:
fix var name

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

mteodor committed:
fix var name

Signed-off-by: Mirko Teodorovic <mirko.teodorovic@gmail.com>

drasko reviewed (approved):
LGTM
